parser/src/parser/grammar.lalrpop

```diff
@@ -57,399 +111,318 @@ Trace: Vec<Vec<TraceBinding>> = {
-    <l:@L> <integrity_stmts_groups: IntegrityStmtGroup+> <r:@R> =>? {
-        let integrity_stmts: Vec<IntegrityStmt> = integrity_stmts_groups.into_iter().flatten().collect();
-        // check if at least one integrity constraint is defined
-        let integrity_constraints_exist = integrity_stmts.iter().any(|stmt| match stmt {
-            IntegrityStmt::Constraint(_) => true,
-            _ => false,
-        });
-        if !integrity_constraints_exist {
-            diagnostics.diagnostic(Severity::Error)
-                .with_message("at least one integrity constraint must be declared")
-                .with_primary_label(span!(l, r), "missing constraint declaration in this section")
-                .emit();
-            return Err(ParseError::Failed.into());
-        }
-        Ok(integrity_stmts)
+StatementBlock: Vec<Statement> = {
+    <Let> => vec![Statement::Let(<>)],
+    <stmts:ConstraintStatement+> <last:Let> => {
+        let mut stmts = stmts;
+        stmts.push(Statement::Let(last));
+        stmts
     },
-    
+    <ConstraintStatement+>,
 }
 
-IntegrityStmtGroup: Vec<IntegrityStmt> = {
-    "let" <name: Identifier> "=" <integrity_variable_type: IntegrityVariableType> =>
-        vec![IntegrityStmt::VariableBinding(VariableBinding::new(name, integrity_variable_type))],
-    "enf" <integrity_constraint: IntegrityConstraintExpr> => vec![integrity_constraint],
-    "match" "enf" ":" <integrity_stmts: IntegrityConstraintExpr+> => integrity_stmts,
+Let: Let = {
+    <l:@L> "let" <name: Identifier> "=" <value: Expr> <r:@R> <body: StatementBlock>
+        => Let::new(span!(l, r), name, value, body)
 }
 
-IntegrityConstraintExpr: IntegrityStmt = {
-    <constraint: IntegrityConstraintExprValue>
-        <comprehension: ConstraintComprehension<IntegrityExpr>?>
-        <selectors: WithSelectors?> =>
-            IntegrityStmt::Constraint(IntegrityConstraint::new(constraint, comprehension, selectors)),
+ConstraintStatement: Statement = {
+    "enf" <ConstraintExpr> => <>,
 }
 
-IntegrityConstraintExprValue: ConstraintExpr = {
-    <lhs: IntegrityExpr> "=" <rhs: IntegrityExpr> =>
-        ConstraintExpr::Inline(InlineConstraintExpr::new(lhs, rhs)),
-    <evaluator_fn_call: EvaluatorFunctionCall> => ConstraintExpr::Evaluator(evaluator_fn_call)
-}
-
-WithSelectors: Expression = {
-    "when" <selectors: IntegrityExpr> => selectors,
+ConstraintExpr: Statement = {
+    <l:@L> <expr: ScalarConstraintExpr> <comprehension: ConstraintComprehension<ScalarExpr>?> <selector: WithSelector?> <r:@R> => {
+        if let Some(context) = comprehension {
+            Statement::EnforceAll(ListComprehension::new(span!(l, r), expr, context, selector))
+        } else {
+            // If a selector is provided for a scalar constraint, we convert it to a comprehension constraint to keep the AST simple
+            //
+            // The generated comprehension is equivalent to `enf <expr> for %0 in 0..1 when <selector>`
```

Comment:
I've reworked this commentary to cover the entire `ConstraintExpr` rule, and to document more completely the different syntax variations which the grammar rule handles, the different constraint "types" which are represented in the AST, and how/why the different variations map to those constraint types. Since that rule is the only place where constraints are constructed by the parser, I think that should cover things pretty well now.